buffer/src/main/java/io/netty/buffer/BufferOutputStream.java

```diff
@@ -0,0 +1,191 @@
+    private DataOutputStream utf8out; // lazily-instantiated
+    private boolean closed;
+
+    /**
+     * Creates a new stream which writes data to the specified {@code buffer}.
+     */
+    public BufferOutputStream(Buffer buffer) {
+        this.buffer = requireNonNull(buffer, "buffer");
+        startIndex = buffer.writerOffset();
+    }
+
+    /**
+     * Returns the number of written bytes by this stream so far.
+     */
+    public int writtenBytes() {
+        return buffer.writerOffset() - startIndex;
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        if (len == 0) {
+            return;
+        }
+
+        prepareWrite(len);
+        buffer.writeBytes(b, off, len);
+    }
+
+    @Override
+    public void write(byte[] b) throws IOException {
+        if (b.length == 0) {
+            return;
+        }
+        prepareWrite(b.length);
+        buffer.writeBytes(b);
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        prepareWrite(Integer.BYTES);
+        buffer.writeByte((byte) b);
+    }
+
+    @Override
+    public void writeBoolean(boolean v) throws IOException {
+        prepareWrite(1);
+        buffer.writeByte((byte) (v ? 1 : 0));
+    }
+
+    @Override
+    public void writeByte(int v) throws IOException {
+        prepareWrite(1);
+        buffer.writeByte((byte) v);
+    }
+
+    @Override
+    public void writeBytes(String s) throws IOException {
+        prepareWrite(s.length()); // US ASCII has no multibyte values, so this works.
+        buffer.writeCharSequence(s, CharsetUtil.US_ASCII);
+    }
+
+    @Override
+    public void writeChar(int v) throws IOException {
+        prepareWrite(Character.BYTES);
+        buffer.writeChar((char) v);
+    }
+
+    @Override
+    public void writeChars(String s) throws IOException {
+        int len = s.length();
+        prepareWrite(len * Character.BYTES);
+        for (int i = 0 ; i < len ; i ++) {
+            buffer.writeChar(s.charAt(i));
+        }
+    }
+
+    @Override
+    public void writeDouble(double v) throws IOException {
+        prepareWrite(Double.BYTES);
+        buffer.writeDouble(v);
+    }
+
+    @Override
+    public void writeFloat(float v) throws IOException {
+        prepareWrite(Float.BYTES);
+        buffer.writeFloat(v);
+    }
+
+    @Override
+    public void writeInt(int v) throws IOException {
+        prepareWrite(Integer.BYTES);
+        buffer.writeInt(v);
+    }
+
+    @Override
+    public void writeLong(long v) throws IOException {
+        prepareWrite(Long.BYTES);
+        buffer.writeLong(v);
+    }
+
+    @Override
+    public void writeShort(int v) throws IOException {
+        prepareWrite(Short.BYTES);
+        buffer.writeShort((short) v);
+    }
+
+    @Override
+    public void writeUTF(String s) throws IOException {
+        DataOutputStream out = utf8out;
+        if (out == null) {
+            if (closed) {
```

Comment:
I think it would be better to just move this out of the if block as the first action of this method